x-pack/plugin/autoscaling/src/main/java/org/elasticsearch/xpack/autoscaling/storage/ReactiveStorageDeciderService.java

```diff
@@ -0,0 +1,365 @@
+import org.elasticsearch.common.settings.Setting;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.ByteSizeUnit;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.snapshots.SnapshotShardSizeInfo;
+import org.elasticsearch.xpack.autoscaling.capacity.AutoscalingCapacity;
+import org.elasticsearch.xpack.autoscaling.capacity.AutoscalingDeciderContext;
+import org.elasticsearch.xpack.autoscaling.capacity.AutoscalingDeciderResult;
+import org.elasticsearch.xpack.autoscaling.capacity.AutoscalingDeciderService;
+import org.elasticsearch.xpack.core.DataTier;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+public class ReactiveStorageDeciderService implements AutoscalingDeciderService {
+    public static final String NAME = "reactive_storage";
+
+    private final DiskThresholdSettings diskThresholdSettings;
+    private final AllocationDeciders allocationDeciders;
+
+    public ReactiveStorageDeciderService(Settings settings, ClusterSettings clusterSettings, AllocationDeciders allocationDeciders) {
+        this.diskThresholdSettings = new DiskThresholdSettings(settings, clusterSettings);
+        this.allocationDeciders = allocationDeciders;
+    }
+
+    @Override
+    public String name() {
+        return NAME;
+    }
+
+    @Override
+    public List<Setting<?>> deciderSettings() {
+        return List.of();
+    }
+
+    @Override
+    public List<DiscoveryNodeRole> roles() {
+        return List.of(
+            DiscoveryNodeRole.DATA_ROLE,
+            DataTier.DATA_CONTENT_NODE_ROLE,
+            DataTier.DATA_HOT_NODE_ROLE,
+            DataTier.DATA_WARM_NODE_ROLE,
+            DataTier.DATA_COLD_NODE_ROLE
+        );
+    }
+
+    @Override
+    public AutoscalingDeciderResult scale(Settings configuration, AutoscalingDeciderContext context) {
+        AutoscalingCapacity autoscalingCapacity = context.currentCapacity();
+        if (autoscalingCapacity == null || autoscalingCapacity.total().storage() == null) {
+            return new AutoscalingDeciderResult(null, new ReactiveReason("current capacity not available", -1, -1));
+        }
+
+        AllocationState allocationState = new AllocationState(context, diskThresholdSettings, allocationDeciders);
+        long unassigned = allocationState.storagePreventsAllocation();
+        long assigned = allocationState.storagePreventsRemainOrMove();
+        long maxShard = allocationState.maxShardSize();
+        assert assigned >= 0;
+        assert unassigned >= 0;
+        assert maxShard >= 0;
+        String message = unassigned > 0 || assigned > 0 ? "not enough storage available, needs " + (unassigned + assigned) : "storage ok";
```

Comment:
I wonder if this should be human readable bytes? So `new ByteSizeValue(unassigned + assigned).toString()`?

Reply:
And if not, ` bytes` should be appended to the message.

Reply:
:+1:, b58d29462a12559caaeef09b7aae0193a4a65ef9